Add SPEC7: Seeding pseudo-random number generation (#180)
scientific-python/specs: spec-0007/index.md

---
title: "SPEC 7 — Seeding pseudo-random number generation"
date: 2023-04-19
author:
  - "Stéfan van der Walt <stefanv@berkeley.edu>"

>   - Other participants in the discussion <not.yet@named.org>"

Reviewer: ```suggestion
  - "Pamphile Roy <roy.pamphile@gmail.com>"
  - "Sebastian Berg <...>"
```

@seberg 😉

---
title: "SPEC 7 — Seeding pseudo-random number generation"
date: 2023-04-19
author:
  - "Stéfan van der Walt <stefanv@berkeley.edu>"
  - "Sebastian Berg <sebastianb@nvidia.com>"
  - "Pamphile Roy <roy.pamphile@gmail.com>"
  - "Matt Haberland <mhaberla@calpoly.edu>"
  - Other participants in the discussion <not.yet@named.org>"
discussion: https://github.com/scipy/scipy/issues/14322
endorsed-by:
---

## Description

Currently, libraries across the ecosystem provide various APIs for seeding random number generation.
This SPEC suggests a unified, pragmatic API, taking into account technical and historical factors.
Adopting such a uniform API will simplify the user experience, especially for those who rely on multiple projects.

We recommend:

- avoiding the use of global state and legacy bitstream generators, and
- standardizing the usage and interpretation of an `rng` keyword for seeding.

We suggest implementing these principles by:

> - deprecating the use of `numpy.random.seed` to control the random state,

Reviewer 2: Maybe should clarify that this recommendation is for downstream libraries, not NumPy itself.
And I suppose in the end we didn't exactly *deprecate* this either : / But I suppose I'm splitting hairs about the word.

- deprecating the use of `random_state`/`seed` arguments in favor of a consistent `rng` argument, and
- using `numpy.random.default_rng` to validate the `rng` argument and instantiate a `Generator`.

> Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.

Reviewer 2: Using `default_rng` to validate `rng` has been described as the ["blessed One True Pattern"](https://github.com/scientific-python/specs/pull/180#discussion_r1667389979).

> Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.

Reviewer 3: 1. We intended to add the coercion of `RandomState` to `default_rng()` (following our experience with SciPy's QMC version of `check_random_state()` which coerces things to `Generator`s) but never got around to it. It's a good idea, but a little bit optional. Once people are passing objects to `rng=` keywords, they probably won't be doing so with `RandomState` objects. But it'll be a good corner case to patch up. This is on NumPy.

2. The use case of `RandomState` inside of unit tests is about calling `RandomState` methods explicitly inside of unit tests to create data arrays. The use case is not passing `RandomState` instances to the functions under test. Those functions will soon all be using `Generator`s inside their internals, obliterating any of the frozen functionality of `RandomState` regardless.

> Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.

Reviewer 4: Good points @mdhaber and @rkern. All makes sense. I suggest to just add a little edit to the text to make this clearer, something like _"is effectively deprecated (as input to `rng=`, not for generating data in unit tests)"_

> Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.

Reviewer 2: Addressed by https://github.com/stefanv/specs/pull/3/commits/37ff883a4b5bd3272e5aea3ce75ab36f851ac8ce

`RandomState` instances via an argument with a name other than `rng`, but the ideas
are more broadly applicable. Use of random number generators other than those provided
by NumPy are beyond the scope of this SPEC.

### Concepts

- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`numpy.random.default_rng`) uses PCG64.
- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.

### Constraints

NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
Common keyword arguments include `random_state` and `seed`.
In practice, the seed is unfortunately also often controlled using `numpy.random.seed`.

## Implementation

Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.

> Two strong motivations for moving over to `Generator`s are:

Reviewer 4: I recommend splitting this off into a separate `## Motivation` section which goes above the Implementation section. That will help the reader, and it will make clearer to the authors how short the motivation given now is (it needs expanding).

Reviewer 2: Can do.

Reviewer 2: Partially addressed by https://github.com/scientific-python/specs/pull/180#discussion_r1705091496, but I leave addition of motivation for a future PR. I'm not the best person to comment on this; I was comfortable controlling using `np.random.rand` and only changed due to peer pressure : )

Reviewer 2: Yup, fixed it.

Author: The SPEC template doesn't have a motivation section (/cc @jarrodmillman perhaps we should consider), but I added some motivational text to the description.

1. they avoid naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
2. they avoid using global state (from `numpy.random.mtrand._rand`).

Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen principle[^hinsen],
although it could very nearly do so by enforcing the use of `rng` as a keyword argument.

[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.

The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:

1. Accept both `rng` and `random_state`/`seed` keyword arguments.
2. If both are specified, raise an error.
3. If neither is specified and `np.random.seed` has been used to set the seed, emit a `FutureWarning` about the upcoming change in behavior.
4. If `random_state`/`seed` is passed by keyword or by position, treat it as before, but:

>    - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.

Reviewer 4: I suggest adding a recommendation here to wait N releases (with N determined from SPEC 0) before starting to emit a deprecation warning after introducing `rng=`, because downstream libraries cannot switch to the new keyword when they're still supporting library versions that don't have it.

Reviewer 2: Would we specify a number of releases or number of years?

Reviewer 4: SPEC 0 is in years, but projects work in releases. So the text could be amended with either option. Here is a concrete suggestion: _"After `rng=` is introduced and is available in all releases within the support window suggested by SPEC 0 for the package, emit a `DeprecationWarning` ...."_

Reviewer 2: That I can do. I just couldn't specify `N`. Will do today.

Reviewer 2: Addressed by https://github.com/stefanv/specs/pull/3/commits/6ec3b6ae1dd5ccc012eb16b5d2c218194c9120fa

Reviewer 4: Thanks. The change LGTM.

   ```python
   from collections.abc import Sequence
   import numpy as np

   SeedLike = int | np.integer | Sequence[int] | np.random.SeedSequence
   RNGLike = np.random.Generator | np.random.BitGenerator

   def my_func(rng: RNGLike | SeedLike | None = None):
       """My function summary.

       Parameters
       ----------
       rng : `numpy.random.Generator`, optional
           Pseudorandom number generator state. When `rng` is None, a new
           `numpy.random.Generator` is created using entropy from the
           operating system. Types other than `numpy.random.Generator` are
           passed to `numpy.random.default_rng` to instantiate a `Generator`.
       """
       rng = np.random.default_rng(rng)

       ...

   ```

   Also note the suggested language for the `rng` parameter docstring, which encourages the user to pass a `Generator` or None, but allows for other types accepted by `numpy.random.default_rng` (captured by the type annotation).

### Impact

There are three classes of users, which will be affected to varying degrees.

1. Those who do not attempt to control the random state.

>    Their code will immediately switch from using the unseeded global `RandomState` to using an unseeded `Generator`.

Reviewer 2: ```suggestion
   Their code will switch from using the unseeded global `RandomState` to using an unseeded `Generator`.
```
There's not really a reason that it needs to be immediate, and that's not what the decorator does now.

1. `Generator`s avoid problems associated with naïve seeding (e.g., using successive integers), via its [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning) mechanism;
2. their use avoids relying on global state—which can make code execution harder to track, and may cause problems in parallel processing scenarios.

[^no-RandomState]:
    Note that `numpy.random.default_rng` does not accept instances of `RandomState`, so use of `RandomState` to control the seed is effectively deprecated, too.
    That said, neither `np.random.seed` nor `np.random.RandomState` _themselves_ are deprecated, so they may still be used in some contexts (e.g. by developers for generating unit test data).

### Scope

This is intended as a recommendation to all libraries that allow users to control the state of a NumPy random number generator.
It is specifically targeted toward functions that currently accept `RandomState` instances via an argument other than `rng`, or allow `numpy.random.seed` to control the random state, but the ideas are more broadly applicable.

> Use of random number generators other than those provided by NumPy are beyond the scope of this SPEC.

Author: I think the API is generic, it's just that we currently advocate that the `rng` argument be sent into `np.random.default_rng`. If you want to accept other types of seeds, or objects other than `Generator`s, you can still use the `rng` keyword, you just have to know that this was not thought through as part of the SPEC.

> Use of random number generators other than those provided by NumPy are beyond the scope of this SPEC.

Author: I left a note.

although it could very nearly do so by enforcing the use of `rng` as a keyword argument.

[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.

The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is as follows.

**Initially**, accept both `rng` and the existing `random_state`/`seed`/`...` keyword arguments.

- If both are specified by the user, raise an error.
- If `rng` is passed by keyword, validate it with `np.random.default_rng()` and use it to generate random numbers as needed.
- If `random_state`/`seed`/`...` is specified (by keyword or position, if allowed), preserve existing behavior.

**After `rng` becomes available** in all releases within the support window suggested by SPEC 0, emit warnings as follows:

- If neither `rng` nor `random_state`/`seed`/`...` is specified and `np.random.seed` has been used to set the seed, emit a `FutureWarning` about the upcoming change in behavior.
- If `random_state`/`seed`/`...` is passed by keyword or by position, treat it as before, but:

>   - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.

Author: I'll let @mdhaber take that one.

>   - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.

Reviewer 2: I was not familiar with that being the official distinction. This was written based on what I've seen in SciPy. 

In our release notes, there is a standard note about *users* checking for `DeprecationWarning`s:

> Before upgrading, we recommend that users check that their own code does not use deprecated SciPy functionality (to do so, run your code with `python -Wd` and check for `DeprecationWarning`s).

In SciPy, I have always seen a `DeprecationWarning` when something is going to raise an error in the future (i.e. in line with Hinsen principle) and a `FutureWarning` when something is going to change behavior but not raise an error (i.e. in violation of the Hinsen principle).

Often `DeprecationWarning`s are used for functions and keywords being removed. `FutureWarning`s are a lot less common. [Here](https://github.com/scipy/scipy/pull/21446#issuecomment-2310197404)'s an example of a decision made between the two.

In this case, a keyword is being removed in the sense that if a user passes `random_state=10` they will get an error, so SciPy would probably emit a `DeprecationWarning`. If they pass `10` to `random_state` by position, on the other hand, they will not get an error in the future, but the behavior will change, so SciPy would probably emit a `FutureWarning`.

I suppose we can mention that projects are welcome to follow their conventions about which type of warning to emit.

>   - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.

Author: The main important distinction here seem to be: by default, users won't *see* the `DeprecationWarning`s (without explicitly hunting for them with `-Wd`), whereas they *will* see `FutureWarning`s "out the box".

>   - Emit a `DeprecationWarning` if passed by keyword, warning about the deprecation of keyword `random_state` in favor of `rng`.

Reviewer 5: Yeah, that's usually why we default to `FutureWarning` in scikit-image. Though, I'm happy for this SPEC to recommend whatever. :wink:

  - Emit a `FutureWarning` if passed by position, warning about the change in behavior of the positional argument.

**After the deprecation period**, accept only `rng`, raising an error if `random_state`/`seed`/`...` is provided.

By now, the function signature, with type annotations, could look like this:

```python
from collections.abc import Sequence
import numpy as np

SeedLike = int | np.integer | Sequence[int] | np.random.SeedSequence
RNGLike = np.random.Generator | np.random.BitGenerator

> def my_func(rng: RNGLike | SeedLike | None = None):

Reviewer 2: BTW, is there a good reference for the name "Hinsen principle"? The idea is great, but this is the Google search result for the "famous" Hinsen principle:

![image](https://github.com/user-attachments/assets/bf24f5f0-5cac-42a3-9bdc-756e8f6ed01c)

> def my_func(rng: RNGLike | SeedLike | None = None):

Reviewer 6: I also had to google for it, but then found a couple of interesting reads that now populates a few browser tabs set aside to be read later :)

> def my_func(rng: RNGLike | SeedLike | None = None):

Author: I suspect it comes from conversations with @khinsen, or perhaps his blog post https://blog.khinsen.net/posts/2017/11/16/a-plea-for-stability-in-the-scipy-ecosystem.html

Konrad complains about the scientific Python ecosystem a fair bit, especially about how we don't pay attention to backward compatibility. So, I suppose it is ironic that we spend so much time on this problem, and only fitting that we name one of our biggest headaches after him ;) 

[Konrad doesn't seem like someone who'd take offense at my humor, but I'll make clear that that's what it is either way.]

> def my_func(rng: RNGLike | SeedLike | None = None):

Reviewer 2: > Unless switching the behavior in-place is actually intended like the decorator implies

The decorator includes as a comment:
```
                # Note that this decorator does not deprecate positional use of the
                # argument; it only warns that the behavior will change in the future.
                # Simultaneously transitioning to keyword-only use is another option.
```
and the SPEC notes:

> Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen principle[^hinsen], although it could very nearly do so by enforcing the use of `rng` as a keyword argument.

So the decorator does not mean to imply that this is the way to go.

When I first started generalizing the decorator, which worked only if `random_state` (or other old name) was already keyword-only, I had in mind that it would make `rng` keyword-only regardless of whether `random_state` was keyword-only or not. Some considerations that changed the plan to continue to allow positional use:

- Even though I think both of them support making arguments keyword-only, Stefan and Pampile mentioned that the SPEC should not be overly-prescriptive about that. I agree that I wouldn't want this to be held up by a debate about the merits of transitioning to keyword-only arguments; it's complicated enough as it is!
- If `random_state` (or other old name) currently allows positional use, deprecating positional use of `random_state` must be accompanied by deprecating positional use of all following positional arguments. That is a bit more of a disruption.
- There are some other complications depending on what we want the documented signature to look like:
  - If we also want the documented signature to change immediately (e.g. so auto-completion suggests the new signature) as we do now, the decorator has to pass arguments to the underlying function by keyword, which means that the decorator has to know their keyword names, requiring either more complex decorator configuration or use of introspection. 
  - If we _don't_ change the documented signature immediately, we need some sort of admonitions in the documentation about the deprecation of the keywords, which is also going to be more complex.

In short, I took what appeared to be the safer route, but projects are welcome to make it keyword-only at the same time if they wish.

> def my_func(rng: RNGLike | SeedLike | None = None):

Reviewer 7: > Konrad complains about the scientific Python ecosystem a fair bit, especially about how we don't pay attention to backward compatibility. So, I suppose it is ironic that we spend so much time on this problem, and only fitting that we name one of our biggest headaches after him ;)

Well, I am an academic, any citation is good for me ;-)

It's just a bit sad that Google considers me a typo for a brand of TV sets. But those are the laws of the market...

> def my_func(rng: RNGLike | SeedLike | None = None):

Author: > Well, I am an academic, any citation is good for me ;-)

:joy: 

Since we have you here, Konrad, would you say the above is the best citation of yours to the principle? I can add the URL to the SPEC and to the skimage discussion, so people will hopefully find it more easily in the future.

Alternatively, we can do something a tiny bit more ambitious and write a blog post at https://blog.scientific-python.org to explain the problem.

I don't know if anyone has been following, but scikit-image has been in a major holding pattern over this for years now. We're trying terribly hard to get a new version of the package out, with API designs we consider superior, but in many cases run into silent backward incompatibility. Our only viable solution so far is to release into an entirely new namespace, `skimage2`, which in itself is a headache. We haven't proceeded with that yet, because we are somewhat terrified of estranging existing users, but at the same time know we have to move forward.

> def my_func(rng: RNGLike | SeedLike | None = None):

Author: I should also probably apologize to @khinsen for using the name Scientific Python without his permission (ref https://pypi.org/project/ScientificPython/). An oversight at the time, but hopefully we will receive forgiveness, given that we're trying to fulfill a mission I believe he is strongly in favor of.

> def my_func(rng: RNGLike | SeedLike | None = None):

Reviewer 5: @mdhaber, I'm not clear on your view on the current change to recommend keyword-only here. Is that okay or would you prefer to revert it before we merge this?

> def my_func(rng: RNGLike | SeedLike | None = None):

Reviewer 2: I think keyword-only is preferable, too. If the original argument was keyword-only, I think the decorator can keep it keyword-only, so this is not necessarily inconsistent with what the decorator can do. 

I'll think about extending it to make it able to deprecate positional use of an argument if desired. But that doesn't need to hold anything up.

> def my_func(rng: RNGLike | SeedLike | None = None):

Reviewer 7: > would you say the above is the best citation of yours to the principle?

Yes!

> Our only viable solution so far is to release into an entirely new namespace, skimage2, which in itself is a headache.

I see why it's a headache, but I also believe it's the best solution in the long run. Explicit is better than implicit. New API, new name. And it's [done elsewhere](http://blog.ezyang.com/2016/12/thoughts-about-spec-ulation-rich-hickey/) with good success.

> hopefully we will receive forgiveness, given that we're trying to fulfill a mission I believe he is strongly in favor of.

Indeed I am! The name clash has regularly been a source of confusion, but nothing serious. And by now, my ScientificPython has disappeared from most people's radar, being Python-2-only.

### Impact

There are three classes of users, which will be affected to varying degrees.

1. Those who do not attempt to control the random state.
   Their code will switch from using the unseeded global `RandomState` to using an unseeded `Generator`.
   Since the underlying _distributions_ of pseudo-random numbers will not change, these users should be unaffected.

2. Users of `random_state`/`seed` arguments.
   Support for these arguments will be dropped eventually, but during the deprecation period, we can provide clear guidance, via warnings and documentation, on how to migrate to the new `rng` keyword.

3. Those who use `numpy.random.seed`.
   The proposal will do away with that global seeding mechanism, meaning that code that relies on it would, after the deprecation period, go from being seeded to being unseeded.
   To ensure that this does not go unnoticed, libraries that allowed for control of the random state via `numpy.random.seed` should raise a `FutureWarning` if `np.random.seed` has been called. (See [Code](#code) below for an example.)

>    In response, users must switch from using `numpy.random.seed` to passing the `rng` argument explicitly to all functions that accept it.

Author: I added two notes to make it clear.

---
title: "SPEC 7 — Seeding pseudo-random number generation"
date: 2023-04-19
author:
  - "Stéfan van der Walt <stefanv@berkeley.edu>"
  - "Sebastian Berg <sebastianb@nvidia.com>"
  - Other participants in the discussion <not.yet@named.org>"
discussion: https://github.com/scipy/scipy/issues/14322
endorsed-by:
---

## Description

<!--
Briefly and clearly describe the proposal.
Explain the general need and the advantages of this specific proposal.
If relevant, include examples of how the new functionality would be used,
intended use-cases, and pseudo-code illustrating its use.
-->

There is disparity in the APIs libraries provide to seed random number generation.
This SPEC suggests a single, pragmatic API for the ecosystem, taking into account technical and historical factors.
Adopting such a uniform API will simplify the user experience, especially for those who rely on multiple projects.

Specifically, we recommend to:

- Deprecate the use of `RandomState` and `np.random.seed`.

> - Standardize usage and interpretation of an `rng` keyword for seeding.

Reviewer 8: I think we can split this into two via something like

```suggestion
- Pass the default or seeded `numpy.random.Generator` instances to functions.
- Standardize the `rng` acronym for the keyword in function signatures.
```

so that the context is separated from the keyword standardization and the actual deprecation. Please adjust the sentences as you see fit.

Author: Thanks for the feedback, @ilayn. Currently, the recommendation is not to follow any specific technical path, but to (a) avoid global state and (b) use a standard mechanism for sending generators around.

Reading your suggestion, I think it could be added as a second point, i.e., write down the principles, followed by how they should be enacted. E.g.:

```
We recommend:

- Standardizing the usage and interpretation of an `rng` keyword for seeding, and
- avoiding the use of global state and legacy bitstream generators.

We suggest implementing these principles by:

- Passing the `rng` argument to `numpy.random.default_rng` to instantiate a `Generator`, and
- deprecating  the use of `RandomState` and `np.seed`.
```

Thoughts?

Reviewer 8: Ah yes just wanted to make sure that implementation does not get mixed up with the SPEC goal. All good from my side.

   Such code will, in effect, go from being seeded to being unseeded.
   To avoid that from happening, the code will have to be modified to pass in explicitly an `rng` argument on each function call.

2. Those who do not seed. Their code will, after the deprecation period, use the newly proposed default. Since they were already not requesting repeatable sequences, and since the underlying _distributions_ of pseudo-random numbers did not change, they should be unaffected.

3. Users of `random_state=...`. Support for the `random_state` argument may be dropped eventually, but meanwhile we can provide clear guidance, via deprecation warnings and documentation, on how to migrate to the new `rng` keyword.

[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.

### Code

> 

Reviewer 2: Is there a recommended wording for how the `rng` parameter should be documented?

Reviewer 3: I do recommend emphasizing in each docstring that we expect `Generator`s or `None` (and a quick word about what `None` means) but then link to one common place in your docs that documents all of the rarer cases like static seeds and legacy `RandomState`s. If you use type annotations, consider defining a type that incorporates all the possibilities accepted by `np.random.default_rng()` and documenting that type definition as the place to link those docstrings to.

Author: I thought it might be useful to add that type annotation as part of this spec. However, NumPy type annotates with hidden types (`numpy._typing`). This feels like a mistake?

E.g., I tried:

```
import numpy as np
import numpy.typing

type Seed = int | np.typing.ArrayLike | np.random.SeedSequence | np.random.BitGenerator | np.random.Generator | None

def foo(rng : Seed = None):
    rng = np.random.default_rng(rng)

    return rng.random((3, 3))
```

But then mypy (main) complains:

```
$ mypy --enable-incomplete-feature=NewGenericSyntax foo.py
foo.py:9: error: Argument 1 to "default_rng" has incompatible type "int | Buffer | _SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes] | SeedSequence | BitGenerator | Generator | None"; expected "_SupportsArray[dtype[bool_ | integer[Any]]] | _NestedSequence[_SupportsArray[dtype[bool_ | integer[Any]]]] | bool | int | _NestedSequence[bool | int] | SeedSequence | BitGenerator | Generator | None"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

Should NumPy not expose type declarations used in the public API for re-use by other projects?

Reviewer 4: +1 for adding type annotations.

That has a syntax error (`type Seed`) and a mistake (`rng =` is wrong). The issue is with array-likes (as usual). This works:
```python
import numpy as np
import numpy.typing

# leaving out np.typing.ArrayLike
Seed = None | int | np.random.SeedSequence | np.random.BitGenerator | np.random.Generator | np.typing.NDArray

def foo(rng : Seed = None):
    rng_ = np.random.default_rng(rng)  # note the renamed variable `rng_`

    return rng_.random((3, 3))
```

> Should NumPy not expose type declarations used in the public API for re-use by other projects?

The type annotation in NumPy for `default_rng` captures the requirement that only integer array-likes are accepted. This is correct and precise, but requires a fairly complicated annotation (look for `_ArrayLikeInt_co` in `numpy/_typing`). It does not make sense to expose such things publicly.

Array-likes in general should not be encouraged to propagate, so I'd choose to leave out `ArrayLike` completely, and replace it with `np.typing.NDArray`. That should also still gain an integer specifier: `np.typing.NDArray[np.integer]` should work.

The NumPy type annotations themselves are aiming for precision and completeness. In a SPEC, I think you want to not do that - allowing everything that numpy itself has to allow for corner-case backwards compatibility doesn't seem useful. Instead, you want to clearly capture what the function actually needs. `NDArray[np.integer]` should do that; jumping through hoops to allow arbitrary sequences of ints or even bools is counterproductive.

Mypy will be happy as long as the annotation you use is a subset of what `default_rng` accepts, rather than a superset.

Reviewer 4: I'd also choose the type alias name for what's preferred - so either `Rng` or something like `RngOrSeed`, but not `Seed.

Reviewer 3: Actually, lists and tuples of arbitrary-sized integers are vastly more common (and preferred) than integer arrays.

We're also missing the legacy `RandomState`, for completeness.

Author: > That has a syntax error (`type Seed`)

Not actually a syntax error, just newer syntax: https://peps.python.org/pep-0695/

Author: OK, so re-reading this thread, I'm wondering: what is the purpose of the suggested type annotation in the SPEC? Is it is to be an exhaustive description of what `default_rng` accepts, or rather to guide your library users in providing the correct arguments? I suspect we want more of the latter?

To make it very straightforward, we could then type annotate with `rng : np.random.Generator`, given that this is the usage we encourage. If we wanted to be a bit more flexible, while still guiding the user in the right direction perhaps one of:

```
type SeedOrGenerator = np.random.Generator | int | None
type SeedOrGenerator = np.random.Generator | np.random.SeedSequence | int | tuple[int] | list[int] | None
```

Or we could try and capture the whole slew of acceptable arguments, including arrays of ints, etc., but that would scarcely provide any usage guidance.

Reviewer 3: FWIW, I think it's helpful to split this up into a `SeedLike` and an `RNGLike`.

```python
from collections.abc import Sequence
from typing import Optional

import numpy as np

# I don't know if an array of integers will typecheck as a `Sequence[int]`, but it's a rarer use case.
SeedLike = int | np.integer | Sequence[int] | np.random.SeedSequence
RNGLike = np.random.Generator | np.random.BitGenerator | np.random.RandomState
RNGOrSeed = RNGLike | SeedLike

def foo(rng: Optional[RNGOrSeed] = None):
    rng = np.random.default_rng(rng)
    ...
```

Author: To answer my own question from earlier: the purpose of the type annotations are to check the *validity* of calls, primarily, so we cannot use the much restricted subset. That we *can* do in the docstrings, though, with some explanatory narrative.

Author: @rkern I think the following will work, handling `RandomState` separately:

```python
from collections.abc import Sequence
from typing import Optional

import numpy as np

SeedLike = int | np.integer | Sequence[int] | np.random.SeedSequence
RNGLike = np.random.Generator | np.random.BitGenerator | np.random.RandomState

def foo(rng: RNGLike | SeedLike | None = None):
    if isinstance(rng, np.random.RandomState):
        rng = np.random.default_rng(rng._bit_generator)

    rng = np.random.default_rng(rng)
    ...

foo(1)
foo([1, 2, 3])
foo((1, 2, 3))
foo(np.random.default_rng(12))
```

Author: Maybe a bit easier to read:

```python
def ensure_rng(rng : SeedLike | RNGLike | None) -> np.random.Generator:
    if isinstance(rng, np.random.RandomState):
        return np.random.default_rng(rng._bit_generator)
    else:
        return np.random.default_rng(rng)

def foo(rng: RNGLike | SeedLike | None = None):
    rng = ensure_rng(rng)
    ...
```

Reviewer 3: Ah, I confused myself. We've talked about making `default_rng()` accept `RandomState`s, but we never actually pushed that in. `<grumble grumble grumble>`

You can omit `RandomState`, then.

Author: This has now been added to the text.

As an example, consider how SciPy would transition from the `seed` to the `rng` keyword using a decorator.
This is implemented using:
1. A `check_random_state` function which normalizes either old (`seed`) or new (`rng`) input to a `Generator` object.
   If neither `seed` nor `rng` was passed but the user has previously called `np.random.seed()`
   this function gives a `FutureWarning` because the behavior will change as noted in
   the Impact section point 1.

> 2. A decorator to deal with the `seed` to `rng` keyword rename.  In future versions, this will deprecate the `seed` keyword. Meanwhile, it ensures that the documentation and auto-completion only advertises the new parameter name.

Reviewer 5: ```suggestion
2. A decorator to deal with the `seed` to `rng` keyword rename.  In future versions, this will deprecate the keyword-only `seed` parameter. Meanwhile, it ensures that the documentation and auto-completion only advertises the new parameter name.
```

And maybe a note that this approach is only correct if the deprecated parameter is keyword-only?

   Delaying the deprecation ensures that downstream users can switch to `rng=` on all supported SciPy versions when the deprecation happens.

```python
_NoValue = object()  # singleton to indicate not explicitly passed

def check_random_state(seed=_NoValue, rng=_NoValue):
    if rng is not _NoValue and seed is not _NoValue:
        raise TypeError("cannot pass both `rng=` and `random_state=` at the same time.")
    if rng is not _NoValue:

>         return np.random.default_rng(rng)

Reviewer 2: IIRC, this will allow `rng` to be an integer, but below, it says 
> takes an `rng` argument which is guaranteed to be a either a Generator or a RandomState

If we want to train users to think about `rng`s and less about `seed`s, it might make sense to start by insisting on a RNG object and not allowing an integer seed.

Reviewer: I don't recall we explicitly discussed this yet, good that you are bringing this up.

I am +1 on not allowing integers in the long run to, as you said, force people to think about what they are doing when they use a rng. Thinking about it, if you want reproducibility you probably want to seed multiple functions so why pass an integer everywhere instead of a Generator object.

Reviewer 3: It was part of the discussion when we designed `np.random.default_rng()`. We wanted to have a replacement for third-party `check_random_state()` functions inside of `np.random` itself. SPEC7 is about coordinating the deprecation strategy for getting people to move from the pattern of `random_state = check_random_state(random_state)` to `rng = np.random.default_rng(rng)`, not really really re-envisioning much else. This pattern from `check_random_state()` (in it's various flavors) has been battle-tested since its original incarnation in `scikit-learn` for over a decade. Another possibility that was considered was having `default_rng()` only take seeds (or `None`) while also including a `check_rng(rng)` that only takes `Generator`s, `None`, and legacy `RandomState`s, but no one then had the appetite for multiple functions like this, given the success of `check_random_state()`. If you really want to relitigate the issue, the proper venue would be an enhancement proposal for numpy, not a SPEC.

In documentation, I do highly encourage emphasizing that we expect people to pass around `Generator`s primarily, but allowing for a seed (of various kinds) or legacy `RandomState`s are permitted for rarer cases. We don't always know what the entrypoint our user will use in an API. While creating a single `Generator` at the top of a program and passing it all the way through everything is frequently the right thing to do, especially in ML, you come across cases where you will need to rerun one thing (like a CV splitter) multiple times and have it do the same thing (possibly to multiple sequences) separate from the main PRNG stream. Allowing for straight static seeds helps with that (though I do admit there is little loss in requiring people to create the `Generator` from the seed separately).

Reviewer: This SPEC is about what we want it to be about. Personally I am not putting much weight in the argument of time nor existing APIs, simply because a lot of past decisions were made with backward compatibility in mind. Here we have a real opportunity for change.

Let's see what others think.

Reviewer 3: I'm not saying there shouldn't be a discussion. I'm saying that we want the One True Pattern to be in numpy, not the examples of a SPEC to be copied with variations in all the libraries. Right now, the blessed One True Pattern is `default_rng()`. If you want a `default_rng()/check_rng()` pair, propose it in numpy.

Reviewer: Precisely, this is a SPEC and not a NEP. i.e. we can deviate from what NumPy says however we want.

Author: Still, I think having a reference implementation of any code feature that ships along with NumPy is probably wise. And no reason we shouldn't align support patterns between NumPy and the SPEC.

Reviewer 3: My point is precisely that such a change *should* be a NEP and not a SPEC. What you are proposing is certainly plausible. But it was considered and rejected (and no, not for backwards compatibility reasons, but on the grounds of exactly the same information that we have available to us right now).

Reviewer 2: OK, I think I understand that integers and such are still accepted because we want to be able to validate `rng` arguments simply by calling `np.random.default_rng(rng)`; we don't want to have additional checks to reject other types. 

It sounds like we agree that `Generators` should be encouraged, though, even if other types are accepted. With that in mind, how about this for suggested documentation:

```
    rng : `numpy.random.Generator`, optional

        Pseudorandom number generator state. For legacy reasons, types
        other than `numpy.random.Generator` are accepted and passed to
        `numpy.random.default_rng` to instantiate a `Generator`.
```

Reviewer 3: I'd drop "For legacy reasons", but yeah, that works. Consider having a short sentence about `None` providing a fresh, independent `Generator`, since this is a main point of difference between `rng=None` and `random_state=None`; it can use the repetition (unlike all the other minutiae which can be left to the `default_rng` reference).

Author: ```
rng : `numpy.random.Generator`, optional

    Pseudorandom number generator state.  When `rng` is None, a new
    `numpy.random.Generator` is created using entropy from the
    operating system. Types other than `numpy.random.Generator` are
    passed to `numpy.random.default_rng` to instantiate a `Generator`. 
```

Author: I included this in the example function docstring.

    def decorator(fun):
        @functools.wraps(fun)
        def wrapper(*args, **kwargs):
            if old_name in kwargs:
                if dep_version:
                    end_version = dep_version.split('.')
                    end_version[1] = str(int(end_version[1]) + 2)
                    end_version = '.'.join(end_version)
                    message = (f"Use of keyword argument `{old_name}` is "
                               f"deprecated and replaced by `{new_name}`.  "
                               f"Support for `{old_name}` will be removed "
                               f"in SciPy {end_version}.")
                    warnings.warn(message, DeprecationWarning, stacklevel=2)
                if new_name in kwargs:
                    message = (f"{fun.__name__}() got multiple values for "
                               f"argument now known as `{new_name}`")
                    raise TypeError(message)

            kwargs[new_name] = check_random_state(
                kwargs.pop(old_name, _NoValue),
                rng=kwargs.pop(new_name, _NoValue)
            )
            return fun(*args, **kwargs)
        return wrapper
    return decorator

> @_prepare_rng("random_state")

Reviewer 5: Note that this decorator uses the deprecated `random_state` while point 2 above talks about `seed` as the deprecated parameter? Is this a typo or am I confused myself?

Reviewer 9: Ah, should align it to `random_state` probably.  SciPy has both, which is where this slip up cames from probably.

def library_function(/, rng=None):
    # The decorated library function takes an `rng` argument which is
    # guaranteed to be a either a Generator or a RandomState.
    # `random_state=` is supported input (the old can be customized).

>     assert isinstance(rng, (np.random.Generator, np.random.RandomState))

Reviewer 5: Just some optional feedback for this example since I recently wrote a blog post on precisely that: [Deprecate positional arguments with placeholders](https://grueter.dev/blog/deprecate-arg-with-placeholder/). That approach has the advantage that the deprecated parameter still shows up in the signature during the deprecation. If wanted I'm happy to come up with an updated version of this snippet. :wink:

Reviewer 5: Oh, I don't mean to hold this up. We can always update after the merge.

Reviewer: That was what we actually wanted 😅 To not show the deprecated usage. There are various opinions there OC. My view as a user is that I don't like to have to go through all the API doc and pay attention to see if one field is deprecated or not, i.e. just give me the recommended way of doing things.

This is either way just a snippet that project can use as a Quickstart and adapt to their own policy.

Reviewer 6: I agree with Pamphile here and try to remove the deprecated parameters from the docstrings as soon as the deprecation happens to help the transition, and help to avoid new usage of deprecated ways.

Author: Any code updates you'd like to see here, then?

Reviewer 6: This code doesn't seem to do anything about the docstrings, so should be fine?

Though a super explicit sentence about signatures and docstrings may be useful, as the code example really goes down to the weeds of how one does deprecations (e.g. there are tons of [deprecation decorators out there](https://docs.astropy.org/en/latest/api/astropy.utils.decorators.deprecated_renamed_argument.html#astropy.utils.decorators.deprecated_renamed_argument), I don't necessarily see why this code snipped is needed in the SPEC). 
Anyway, I still miss something along the lines of:

-  `random_seed` should be removed from signatures and docstrings in favour of `rng`.

Reviewer 3: I think that's a step beyond the scope of what a SPEC should attempt do.

Author: @rkern Do you mean you think it's best to not give a code example?

Reviewer 6: btw, my comments should not block this to be merged as draft as is.

Reviewer 3: I mean that it's best not to recommend what should/should not be documented.

Reviewer 5: Having read the above, I'd say it's fine to include a snippet. However, it should make it very clear in which situations this can be used.

To add a bit of context to my suggestion. I think this example assumes that the deprecated parameter is keyword-only. For libraries that expose the deprecated `random_state`/`seed` as a positional parameter, this snippet wouldn't help / handle that correctly.

Hence my suggestion above to use a placeholder. Without, the signature doesn't reflect the behavior. Very confusing if a user tries to target a positional parameter after the intended input is swallowed by the deprecated one. That limitation should be addressed either as a comment or by patching the snippet?

Reviewer: I would just be clear at the beginning that this snippet is, well a snippet. i.e. just an aid to help a maintainer quick-start things on their repo.

Author: @mdhaber Is reviewing the text surrounding the snippet now.

- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.

### Constraints

NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
Common keyword arguments include `random_state` and `seed`.
In practice, the seed is unfortunately also often controlled using `np.random.seed`.

## Implementation

<!--
Discuss how this would be implemented.
-->

The new API takes into account legacy behavior in packages such as scikit-learn (see `sklearn.utils.check_random_state`), which works as follows:

1. Because `np.random.seed` is so often used in practice, no seed means
   using the global `RandomState` object, `np.random.mtrand._rand`.
2. (Option a) When a seed is provided, a `RandomState` object is initialized with that seed.
3. (Option b) When a seed is provided, a `Generator` object is initialized with that seed.
4. If an instance of `RandomState` is provided, it is used as-is.

> 5. If an instance of `Generator` is provided, it is used as-is.

Reviewer: This is describing the current state in some libraries. But is it where we want to see this 10 years from now?

I am personally against any global state and advertising of any "legacy" behaviours.

Author: I also feel we may want to think about a new keyword argument instead, that adopts recommended best practices instead.

Reviewer: I am not sure I see how that could help.

If you add a `rng` to a function which has `seed` or `random_state` we don't avoid raising some warning about deprecation, etc.

Author: The big problem with `random_state` is that it allows for `None`, which then grabs global state. So, that will always conflict with an `rng=None` kwarg.

Reviewer: To me, the ideal API (using todays tool at least) would be that `random_state=None` would give you `np.random.default_rng()`.

There is also the crazy thought, which I kind of like, from @ilayn: do not accept integers, only a `Generator` (or other object). Point being, you must provide a RNG if you want any reproducibility.

Author: I don't think we can do that, though, because it would be a backward incompatible change?

Reviewer: Both integer and global state behaviour are BC issues yep.

It would be interesting to see how in real life `Generator` was painful to move to, taking into account a large sample of projects, folks, etc.

Author: Yes, and don't expect the same favors from commercial code. It's shackles that we choose to put on, sometimes at great time cost to  (oftentimes volunteer) developers.

The goal of this SPEC is suggest a single, pragmatic API for the ecosystem, taking into account technical and historical factors.

### Concepts

- `BitGenerator`: Generates a stream of pseudo-random bits. The default generator in NumPy (`np.random.default_rng`) uses PCG64.
- `Generator`: Derives pseudo-random numbers from the bits produced by a `BitGenerator`.
- `RandomState`: a [legacy object in NumPy](https://numpy.org/doc/stable/reference/random/index.html), similar to `Generator`, that produces random numbers based on the Mersenne Twister.

### Constraints

NumPy, SciPy, scikit-learn, scikit-image, and NetworkX all implement pseudo-random seeding in slightly different ways.
Common keyword arguments include `random_state` and `seed`.
In practice, the seed is unfortunately also often controlled using `np.random.seed`.

## Implementation

<!--
Discuss how this would be implemented.
-->

Legacy behavior in packages such as scikit-learn (`sklearn.utils.check_random_state`) typically handle `None` (use the global seed state), an int (convert to `RandomState`), or `RandomState` object.

Two strong motivations for moving over to `Generator`s are:

> 
> (1) it avoids naïve seeding strategies, such as using successive integers, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
> (2) it avoids using global state for seeding.

Reviewer 3: ```suggestion

(1) it allows safe and powerful parallel streams, via the underlying [SeedSequence](https://numpy.org/doc/stable/reference/random/parallel.html#seedsequence-spawning);
(2) it avoids creating implicit global state.
```

Not sure how to rewrite (2) in more positive wording. "for seeding" was a distractor there. We're really referring to the implicit global state of `np.random.mtrand._rand` that is used by the `np.random.random()` et al. convenience functions, and the benefit of the new framework is really that we just don't do that anymore. It fundamentally _is_ something that we "don't do" anymore rather than something that we "do better".

Our recommendation here is a deprecation strategy which does not in _all_ cases adhere to the Hinsen[^hinsen] principle.

The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#issuecomment-1515248009) is:

1. Accept both `rng` and `random_state` keyword arguments.
2. If `rng=None`, handle `random_state` as in legacy behavior (see above), except use a compatible Generator instead of RandomState.
   A DeprecationWarning is raised to warn about a future change in behavior.
3. After <X time>, use only `rng`, seeding with `default_rng(rng)`.
   Raise an error if `random_state` is provided.
4. At a time of the lirbrary's choosing, remove any machinery related to `random_state`.

### Impact

The following users will be affected:

1. Those who use `np.random.seed`. The proposal will do away with that global seeding mechanism, meaning that code that relies on it will, after a certain deprecation period, start seeing a different stream of random numbers than before.

   Such code will, in effect, go from being seeded to being unseeded.

>    To avoid that from happening, the code will have to be modified to pass in explicitly an `rng` argument on each function call.

Reviewer 9: This change should give a `FutureWarning` if the user called `np.random.seed()` at any time.  If they did not call `np.random.seed()` then the warning is unnecessary because the stream is not reproducible.

Author: OK, if NumPy can commit to that, that'd be great! I've added it in.

Author: I think I misunderstood what you meant; you meant the decorator should do this check and warn?

Reviewer 9: Yes, sorry, I tried to fix it a bit in the PR (will update now for the other changes).  However, that code does nevertheless reach into NumPy internals, so there is some amount of "commitment" to not break it.

> 2. Those who do not seed. These users _probably_ care more about _distributions_ of pseudo-random numbers, rather than the numbers themselves. But, because they do not provide an explicit seed, their code will, after the deprecation period, use the newly proposed default.

Reviewer 10: For those that do not seed, their code is expected to have different results. I do not see a big impact here.

3. Users of `random_state=...`. Support for the `random_state` argument may be dropped eventually, but meanwhile we can provide clear guidance, via deprecation warnings and documentation, on how to migrate to the new `rng` keyword.

[^hinsen]: The Hinsen principle states, loosely, that code should, whether executed now or in the future, return the same result, or raise an error.

### Code

> TODO: Add example `check_random_state` implementation.

Reviewer 9: **Suggested edit (since triple ticks won't work anyway)**

For example, SciPy may implement this with the following decorator to support both names.
This is implemented using:
1. A `check_random_state` function which normalizes the either old or new input to a `Generator` or `RandomState` object.
   This function only gives a `FutureWarning` when users (probably) NumPy's global legacy `RandomState`.  As noted in point 1. above.
2. A decorator to deal with the rename changes.  In future versions this will deprecate the old keyword.  At this time, it will ensure that the documentation and auto-completion only shows the new parameter name.
   Delaying the deprecation ensures that downstream users can switch to `rng=` on all supported SciPy versions when the deprecation happens.

```python
_NoValue = object()  # singleton to indicate not explicitly passed

def check_random_state(seed=_NoValue, rng=_NoValue):
    if rng is not _NoValue and seed is not _NoValue:
        raise TypeError("cannot pass both `rng=` and `random_state=` at the same time.")
    if rng is not _NoValue:
        return np.random.default_rng(rng)

    if seed is _NoValue:
        # If the user passed nothing, we have to reach into NumPy here:
        # 1. If np.random.seed(None) was called (or never called), then we can
        #    just use the default_rng (the result is random anyway).
        # 2. If it was called, we must return the global random state object
        #    and warn about future ignoring of seed!
        if np.random.mtrand._rand._bit_generator._seed_seq is not None:
            # The user did not seed, so no need to warn.
            return np.random.default_rng()
        warnings.warn(
            "The NumPy global rng was seeded in call to np.random.seed() "
            "in the future this function will ignore this seed and return "
            "random values as if a new `np.random.default_rng()` was created.",
            FutureWarning, stacklevel=5)
        return np.random.mtrand._rand
    if seed is None or seed is np.random:
        return np.random.mtrand._rand
    if isinstance(seed, (numbers.Integral, np.integer)):
        return np.random.RandomState(seed)
    if isinstance(seed, (np.random.RandomState, np.random.Generator)):
        return seed

    raise ValueError(f"'{seed}' cannot be used to seed a numpy.random.RandomState"
                     " instance")

def _prepare_rng(old_name, dep_version=None):
    new_name = "rng"

    def decorator(fun):
        @functools.wraps(fun)
        def wrapper(*args, **kwargs):
            if old_name in kwargs:
                if dep_version:
                    end_version = dep_version.split('.')
                    end_version[1] = str(int(end_version[1]) + 2)
                    end_version = '.'.join(end_version)
                    message = (f"Use of keyword argument `{old_name}` is "
                               f"deprecated and replaced by `{new_name}`.  "
                               f"Support for `{old_name}` will be removed "
                               f"in SciPy {end_version}.")
                    warnings.warn(message, DeprecationWarning, stacklevel=2)
                if new_name in kwargs:
                    message = (f"{fun.__name__}() got multiple values for "
                               f"argument now known as `{new_name}`")
                    raise TypeError(message)

            kwargs[new_name] = check_random_state(
                kwargs.pop(old_name, _NoValue),
                rng=kwargs.pop(new_name, _NoValue)
            )
            return fun(*args, **kwargs)
        return wrapper
    return decorator
```

Reviewer: @thomasjpfan here says it's good 😅

Author: ```
    if seed is None or seed is np.random:
        return np.random.mtrand._rand
```

What does it mean for `seed` to be `np.random`?

Would be good to also show an example of how to apply the decorator.

Reviewer: > Would be good to also show an example of how to apply the decorator.

Sebastian made a PR to show that in SciPy actually 😃

Reviewer 9: Resolving this, I added it to https://github.com/stefanv/specs/pull/1 if you can merge this.

Reviewer 11: > Suggested edit (since triple ticks won't work anyway)

FYI you *can* nest triple-quoted sections by adding another backtick to the outer section. For example:
````suggestion
```python
_NoValue = object()  # singleton to indicate not explicitly passed
```
````
(the `suggestion` block has 4 backticks)
